### ISSUE TITLE
Prefetching queries

### ISSUE THREAD

**Issue by vdjurdjevic** [nandorojo/swr-firestore#33]
API for prefetching outside of rendering would be awesome. For example, if I have some critical data that makes no sense to show anything before it loads, I would load it before calling `ReactDOM.render`. RactQuery has support for this, take a look.

```typescript
import { queryCache } from 'react-query'
 
const prefetchTodos = async () => {
    const queryData = await queryCache.prefetchQuery('todos', () => fetch('/todos'))
    // The results of this query will be cached like a normal query
}
```

**Comment by nandorojo** [comment]
Could you give a real-world example with this lib?

**Comment by vdjurdjevic** [comment]
Working on it right now. Client-side rendered application that shows splash screen while loading data on the first load. The splash screen is inlined in index.html, and when data is ready, it's removed from the DOM, and ReactDOM.render is called.

**Comment by nandorojo** [comment]
For `getInitialProps` and other Next.js queries, I recommend looking at the examples on this issue: https://github.com/nandorojo/swr-firestore/issues/17#issuecomment-657514602

**Comment by nandorojo** [comment]
Why not just use the hook at the top level of your react component, and use React to render a splash screen?

**Comment by vdjurdjevic** [comment]
Because that way you have to wait for react to download and boot everything before you can display a splash screen. This way, if for example, the user is on a slow 3g network and low-end device accessing the first time or after clearing cache, will get splash screen immediately (or with slight fade-in delay). This can take up to 15 seconds for the smallest app.

**Comment by vdjurdjevic** [comment]
This is how I do it right now, but it would be better if there was API exposed for it:

```typescript
import * as firebase from 'firebase/app';
import 'firebase/firestore';

import { cache } from 'swr';

export async function prefetchCollection(name: string) {
    const data = await firebase.firestore().collection(name).get();
    data.docs.forEach(doc => cache.set(doc.ref.path, doc.data(), false));
}

```

**Comment by nandorojo** [comment]
Got it, that seems reasonable. That would require abstracting [these lines](https://github.com/nandorojo/swr-firestore/blob/e7afd1a554754f4434d857da53e8b2e3bd98c5da/src/hooks/use-swr-collection.ts#L299-L332) into a `getCollection` function outside of the hook, exporting that function, and then calling it from within the hook.

```ts
// src/use-swr-collection.ts

export const getCollection = async <Doc extends Document = Document>(path: string, queryString: string) => { 
  const query: Ref = JSON.parse(queryString) ?? {}
  const ref = createRef(path, query)
  const data: Doc[] = await ref.get().then(querySnapshot => {
    const array: typeof data = []
    querySnapshot.forEach(doc => {
      const docData =
        doc.data({
          serverTimestamps: 'estimate',
        }) ?? empty.object
      const docToAdd = withDocumentDatesParsed(
        {
          ...docData,
          id: doc.id,
          exists: doc.exists,
          hasPendingWrites: doc.metadata.hasPendingWrites,
        } as any,
        dateParser.current
      )
      // update individual docs in the cache
      mutateStatic(`${path}/${doc.id}`, docToAdd, false)
      if (
        isDev &&
        // @ts-ignore
        (docData.exists || docData.id || docData.hasPendingWrites)
      ) {
        console.warn(
          '[use-document] warning: Your document, ',
          doc.id,
          ' is using one of the following reserved fields: [exists, id, hasPendingWrites]. These fields are reserved. Please remove them from your documents.'
        )
      }
      array.push(docToAdd)
    })
    return array
  })
}

...

// in useCollection()
const swr = useSWR<Doc[] | null>(
    // if the path is null, this means we don't want to fetch yet.
    path === null ? null : [path, memoQueryString],
    async (path: string, queryString: string) => {
      if (shouldListen.current) {
        if (unsubscribeRef.current) {
          unsubscribeRef.current()
          unsubscribeRef.current = null
        }
        const { unsubscribe, initialData } = await createListenerAsync<Doc>(
          path,
          queryString,
          dateParser.current
        )
        unsubscribeRef.current = unsubscribe
        return initialData
      }

     // here we call the new function
      const data = await getDocument<Doc>(path, queryString)
      return data
    },
    swrOptions
  )
```

I should be able to do that for the next update.

**Comment by vdjurdjevic** [comment]
Awesome :) Can you do it for single documents also?

**Comment by nandorojo** [comment]
Yup. Would also accept a PR.

**Comment by vdjurdjevic** [comment]
Ok, I can't do it right now, I have really tight deadline with this project, but I will let you know if something changes. Definitely would like to contribute to this project :)

**Comment by nandorojo** [comment]
This is now available in `0.11.2`.

```sh
yarn add @nandorojo/swr-firestore@latest
```

**Comment by vdjurdjevic** [comment]
@nandorojo Great job! Thanks :)
